src/org/exist/source/URLSource.java

```diff
@@ -120,6 +126,26 @@ public Validity isValid(final Source other) {
 		return Validity.INVALID;
 	}
 
+	public Charset getEncoding() throws IOException {
+		if (connection == null) {
+			connection = url.openConnection();
+		}
+		final String contentType = connection.getContentType();
+		if (contentType != null) {
+			final Pattern pattern = PatternFactory.getInstance().getPattern("(?i)\\bcharset=\\s*\"?([^\\s;\"]*)");
+			final Matcher matcher = pattern.matcher(contentType);
+			if (matcher.find()) {
+				try {
+					return Charset.forName(matcher.group(1).trim().toUpperCase());
+				} catch (IllegalArgumentException e) {
+					// unknown or illegal charset
+					return null;
```

Comment:
The HTTP 1.1 spec specifies that if the `charset` parameter is not present and the media type is of a `text` subtype, then the client should assume `ISO-8859-1`, see: https://tools.ietf.org/html/rfc2616#section-3.7.1